File: frontend/src/types/infra_investigation.ts
// ─── Infra Investigation Status ─────────────────────────────────────────────

export type InfraTargetType = "url" | "domain" | "ip" | "hash" | "email";

export type InfraStatus =
  | "pending"
  | "running"
  | "completed"
  | "failed"
  | "stopped";

// ─── Pipeline Stage ──────────────────────────────────────────────────────────

export interface InfraPipelineStage {
  key: string;
  name: string;
  status: "pending" | "running" | "completed" | "failed" | "skipped";
  error?: string | null;
}

// ─── Reputation Results ───────────────────────────────────────────────────────

export interface AbuseIPDBResult {
  is_public: boolean;
  abuse_confidence_score: number;
  country_code: string;
  isp: string;
  domain: string;
  total_reports: number;
  last_reported_at: string | null;
  error?: string;
}

export interface URLhausResult {
  query_status: string; // "is_listed" | "no_results"
  urlhaus_reference: string | null;
  blacklists?: Record<string, string>;
  urls_on_this_host?: Array<{
    url: string;
    url_status: string;
    threat: string;
    date_added: string;
  }>;
  error?: string;
}

export interface ThreatFoxResult {
  query_status: string; // "ok" | "no_results"
  iocs?: Array<{
    ioc: string;
    ioc_type: string;
    threat_type: string;
    malware: string;
    malware_printable: string;
    confidence_level: number;
    first_seen: string;
    last_seen: string;
  }>;
  error?: string;
}

export interface OTXPulsesResult {
  pulse_count: number;
  pulses: Array<{
    name: string;
    description: string;
    tags: string[];
    malware_families: string[];
    targeted_countries: string[];
  }>;
  error?: string;
}

// ─── Enrichment Results ───────────────────────────────────────────────────────

export interface DNSRecord {
  type: string; // "A" | "MX" | "NS" | "TXT" | "CNAME"
  value: string;
  ttl?: number;
}

export interface DNSResult {
  records: DNSRecord[];
  error?: string;
}

export interface WHOISResult {
  registrar: string | null;
  registrant_org: string | null;
  creation_date: string | null;
  expiration_date: string | null;
  updated_date: string | null;
  domain_age_days: number | null;
  is_newly_registered: boolean;
  status: string[];
  error?: string;
}

export interface SSLCertResult {
  subject_cn: string | null;
  issuer_cn: string | null;
  issuer_org: string | null;
  serial_number: string | null;
  not_before: string | null;
  not_after: string | null;
  is_expired: boolean;
  is_self_signed: boolean;
  san_domains: string[];
  error?: string;
}

export interface GeoIPResult {
  ip: string;
  country: string | null;
  country_code: string | null;
  region: string | null;
  city: string | null;
  latitude: number | null;
  longitude: number | null;
  org: string | null;  // ISP / Org name
  asn: string | null;
  timezone: string | null;
  error?: string;
}

// ─── Passive DNS ──────────────────────────────────────────────────────────────

export interface PassiveDNSEntry {
  hostname: string;
  address: string;
  first: string;
  last: string;
  asn: string | null;
  country_code: string | null;
}

export interface PassiveDNSResult {
  passive_dns: PassiveDNSEntry[];
  count: number;
  error?: string;
}

// ─── Threat Indicators ────────────────────────────────────────────────────────

export interface ThreatIndicatorCheck {
  id: string;
  name: string;
  description: string;
  triggered: boolean;
  severity: "low" | "medium" | "high" | "critical" | "info";
  detail?: string;
}

export interface ThreatIndicatorsResult {
  checks: ThreatIndicatorCheck[];
  phishing_score: number;
  total_triggered: number;
}

// ─── Correlation ──────────────────────────────────────────────────────────────

export interface InfraCorrelationRule {
  rule_id: string;
  rule_name: string;
  triggered: boolean;
  confidence: "low" | "medium" | "high";
  description: string;
  evidence: string[];
  relationship_type: string;
}

export interface InfraCorrelationResult {
  rules_evaluated: number;
  rules_triggered: number;
  relationships: InfraCorrelationRule[];
  overall_confidence: "low" | "medium" | "high";
}

// ─── Risk Score ───────────────────────────────────────────────────────────────

export interface InfraRiskBreakdown {
  reputation_score: number;
  infrastructure_score: number;
  phishing_score: number;
  weighted_total: number;
  risk_label: "Clean" | "Low" | "Medium" | "High" | "Critical";
  contributing_factors: string[];
}

// ─── Graph Nodes & Edges ─────────────────────────────────────────────────────

export interface GraphNode {
  id: string;
  label: string;
  type: "target" | "ip" | "domain" | "asn" | "registrar" | "campaign" | "malware";
  risk_level: "clean" | "low" | "medium" | "high" | "critical";
  metadata?: Record<string, string | number | boolean | null>;
}

export interface GraphEdge {
  source: string;
  target: string;
  relationship: string;
  confidence: "low" | "medium" | "high";
}

export interface InfraGraph {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

// ─── AI Summary ───────────────────────────────────────────────────────────────

export interface InfraAISummary {
  executive_summary: string;
  threat_classification: string;
  why_suspicious: string;
  recommended_actions: string[];
  confidence: number;
  error?: string;
}

// ─── Full Results Payload (stored in Supabase `results` JSONB) ────────────────

export interface InfraInvestigationResults {
  target: string;
  target_type: InfraTargetType;
  normalized_target: string;
  // Stage results
  reputation?: {
    abuseipdb?: AbuseIPDBResult;
    urlhaus?: URLhausResult;
    threatfox?: ThreatFoxResult;
    otx?: OTXPulsesResult;
  };
  enrichment?: {
    dns?: DNSResult;
    whois?: WHOISResult;
    ssl?: SSLCertResult;
    geoip?: GeoIPResult;
  };
  passive_dns?: PassiveDNSResult;
  threat_indicators?: ThreatIndicatorsResult;
  correlation?: InfraCorrelationResult;
  risk?: InfraRiskBreakdown;
  graph?: InfraGraph;
  ai_summary?: InfraAISummary;
}

// ─── Main Investigation Model ─────────────────────────────────────────────────

export interface InfraInvestigation {
  id: string;
  target: string;
  target_type: InfraTargetType;
  status: InfraStatus;
  current_stage: string;
  progress_percent: number;
  risk_score: number;
  started_at: string;
  completed_at: string | null;
  results: InfraInvestigationResults | null;
}

// ─── API Response Shapes ──────────────────────────────────────────────────────

export interface InfraInvestigationListItem {
  id: string;
  target: string;
  target_type: InfraTargetType;
  status: InfraStatus;
  current_stage: string;
  progress_percent: number;
  risk_score: number;
  started_at: string;
  completed_at: string | null;
}

export interface InfraCreateRequest {
  target: string;
  enable_passive_dns?: boolean;
  enable_ai_summary?: boolean;
}

// ─── Live Event (SOC Feed) ────────────────────────────────────────────────────

export interface InfraLiveEvent {
  id: string;
  timestamp: string;
  stage: string;
  message: string;
  severity: "info" | "success" | "warning" | "critical";
}

File: backend/app/schemas/infra_investigation.py
"""
Pydantic schemas for the Threat Infrastructure Intelligence Flow.
Mirrors frontend types in src/types/infra_investigation.ts exactly.
"""
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict


# ─── Request ─────────────────────────────────────────────────────────────────

class InfraCreateRequest(BaseModel):
    target: str
    enable_passive_dns: bool = True
    enable_ai_summary: bool = True


# ─── Reputation Results ───────────────────────────────────────────────────────

class AbuseIPDBResult(BaseModel):
    is_public: bool = False
    abuse_confidence_score: int = 0
    country_code: str = ""
    isp: str = ""
    domain: str = ""
    total_reports: int = 0
    last_reported_at: Optional[str] = None
    error: Optional[str] = None


class URLhausResult(BaseModel):
    query_status: str = "no_results"
    urlhaus_reference: Optional[str] = None
    blacklists: Optional[Dict[str, str]] = None
    urls_on_this_host: Optional[List[Dict[str, Any]]] = None
    error: Optional[str] = None


class ThreatFoxResult(BaseModel):
    query_status: str = "no_results"
    iocs: Optional[List[Dict[str, Any]]] = None
    error: Optional[str] = None


class OTXPulsesResult(BaseModel):
    pulse_count: int = 0
    pulses: List[Dict[str, Any]] = []
    error: Optional[str] = None


class ReputationResults(BaseModel):
    abuseipdb: Optional[AbuseIPDBResult] = None
    urlhaus: Optional[URLhausResult] = None
    threatfox: Optional[ThreatFoxResult] = None
    otx: Optional[OTXPulsesResult] = None


# ─── Enrichment Results ───────────────────────────────────────────────────────

class DNSRecord(BaseModel):
    type: str
    value: str
    ttl: Optional[int] = None


class DNSResult(BaseModel):
    records: List[DNSRecord] = []
    error: Optional[str] = None


class WHOISResult(BaseModel):
    registrar: Optional[str] = None
    registrant_org: Optional[str] = None
    creation_date: Optional[str] = None
    expiration_date: Optional[str] = None
    updated_date: Optional[str] = None
    domain_age_days: Optional[int] = None
    is_newly_registered: bool = False
    status: List[str] = []
    error: Optional[str] = None


class SSLCertResult(BaseModel):
    subject_cn: Optional[str] = None
    issuer_cn: Optional[str] = None
    issuer_org: Optional[str] = None
    serial_number: Optional[str] = None
    not_before: Optional[str] = None
    not_after: Optional[str] = None
    is_expired: bool = False
    is_self_signed: bool = False
    san_domains: List[str] = []
    error: Optional[str] = None


class GeoIPResult(BaseModel):
    ip: str = ""
    country: Optional[str] = None
    country_code: Optional[str] = None
    region: Optional[str] = None
    city: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    org: Optional[str] = None
    asn: Optional[str] = None
    timezone: Optional[str] = None
    error: Optional[str] = None


class EnrichmentResults(BaseModel):
    dns: Optional[DNSResult] = None
    whois: Optional[WHOISResult] = None
    ssl: Optional[SSLCertResult] = None
    geoip: Optional[GeoIPResult] = None


# ─── Passive DNS ──────────────────────────────────────────────────────────────

class PassiveDNSEntry(BaseModel):
    hostname: str
    address: str
    first: str
    last: str
    asn: Optional[str] = None
    country_code: Optional[str] = None


class PassiveDNSResult(BaseModel):
    passive_dns: List[PassiveDNSEntry] = []
    count: int = 0
    error: Optional[str] = None


# ─── Threat Indicators ────────────────────────────────────────────────────────

class ThreatIndicatorCheck(BaseModel):
    id: str
    name: str
    description: str
    triggered: bool
    severity: str  # low | medium | high | critical | info
    detail: Optional[str] = None


class ThreatIndicatorsResult(BaseModel):
    checks: List[ThreatIndicatorCheck] = []
    phishing_score: float = 0.0
    total_triggered: int = 0


# ─── Correlation ──────────────────────────────────────────────────────────────

class InfraCorrelationRule(BaseModel):
    rule_id: str
    rule_name: str
    triggered: bool
    confidence: str  # low | medium | high
    description: str
    evidence: List[str] = []
    relationship_type: str


class InfraCorrelationResult(BaseModel):
    rules_evaluated: int = 0
    rules_triggered: int = 0
    relationships: List[InfraCorrelationRule] = []
    overall_confidence: str = "low"


# ─── Risk Score ───────────────────────────────────────────────────────────────

class InfraRiskBreakdown(BaseModel):
    reputation_score: float = 0.0
    infrastructure_score: float = 0.0
    phishing_score: float = 0.0
    weighted_total: float = 0.0
    risk_label: str = "Clean"  # Clean | Low | Medium | High | Critical
    contributing_factors: List[str] = []


# ─── Graph ───────────────────────────────────────────────────────────────────

class GraphNode(BaseModel):
    id: str
    label: str
    type: str  # target | ip | domain | asn | registrar | campaign | malware
    risk_level: str  # clean | low | medium | high | critical
    metadata: Optional[Dict[str, Any]] = None


class GraphEdge(BaseModel):
    source: str
    target: str
    relationship: str
    confidence: str  # low | medium | high


class InfraGraph(BaseModel):
    nodes: List[GraphNode] = []
    edges: List[GraphEdge] = []


# ─── AI Summary ───────────────────────────────────────────────────────────────

class InfraAISummary(BaseModel):
    executive_summary: str = ""
    threat_classification: str = ""
    why_suspicious: str = ""
    recommended_actions: List[str] = []
    confidence: float = 0.0
    error: Optional[str] = None


# ─── Full Results Payload ─────────────────────────────────────────────────────

class InfraInvestigationResults(BaseModel):
    target: str
    target_type: str
    normalized_target: str
    reputation: Optional[ReputationResults] = None
    enrichment: Optional[EnrichmentResults] = None
    passive_dns: Optional[PassiveDNSResult] = None
    threat_indicators: Optional[ThreatIndicatorsResult] = None
    correlation: Optional[InfraCorrelationResult] = None
    risk: Optional[InfraRiskBreakdown] = None
    graph: Optional[InfraGraph] = None
    ai_summary: Optional[InfraAISummary] = None


# ─── API Response Models ──────────────────────────────────────────────────────

class InfraInvestigationListItem(BaseModel):
    id: str
    target: str
    target_type: str
    status: str
    current_stage: str
    progress_percent: float
    risk_score: float
    started_at: datetime
    completed_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class InfraInvestigationDetail(BaseModel):
    id: str
    target: str
    target_type: str
    status: str
    current_stage: str
    progress_percent: float
    risk_score: float
    started_at: datetime
    completed_at: Optional[datetime] = None
    results: Optional[InfraInvestigationResults] = None
    error: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)
